eclair-core/src/main/scala/fr/acinq/eclair/router/BalanceEstimate.scala

```diff
@@ -0,0 +1,170 @@
+ * @param high          upper bound on the balance
+ * @param highTimestamp time at which the upper bound was known to be correct
+ * @param totalCapacity total capacity of all the channels between the pair of nodes
+ * @param halfLife      time after which the certainty of the lower/upper bounds is halved
+ */
+case class BalanceEstimate private(low: MilliSatoshi, lowTimestamp: TimestampSecond, high: MilliSatoshi, highTimestamp: TimestampSecond, totalCapacity: Satoshi, halfLife: FiniteDuration) {
+
+  def otherSide: BalanceEstimate = BalanceEstimate(totalCapacity - high, highTimestamp, totalCapacity - low, lowTimestamp, totalCapacity, halfLife)
+
+  /* When we probe an edge, we get certain information on its balance, we know for sure that it can relay at least X or
+   * at most Y. However after some time has passed, other payments may have changed the balance and we can't be so sure
+   * anymore.
+   * We model this decay with a half-life H: every H units of time, our confidence decreases by half and our estimated
+   * probability distribution gets closer to the baseline uniform distribution of balances between 0 and totalCapacity.
+   */
+  private def decay(amount: MilliSatoshi, successProbabilityAtT: Double, t: TimestampSecond): Double = {
+    val decayRatio = 1 / math.pow(2, (TimestampSecond.now() - t) / halfLife)
+    val baseline = 1 - amount.toLong.toDouble / toMilliSatoshi(totalCapacity).toLong.toDouble
+    baseline * (1 - decayRatio) + successProbabilityAtT * decayRatio
+  }
+
+  def couldNotSend(amount: MilliSatoshi, timestamp: TimestampSecond): BalanceEstimate =
+    if (amount < high) {
+      if (amount > low) {
+        copy(high = amount, highTimestamp = timestamp)
+      } else {
+        // the balance is actually below `low`, we discard our lower bound
+        copy(low = MilliSatoshi(0), lowTimestamp = timestamp, high = amount, highTimestamp = timestamp)
+      }
+    } else {
+      /* We already expected not to be able to relay that amount as it above our upper bound. However if the upper bound
+       * was old enough that replacing it with the current amount decreases the success probability for `high`, then we
+       * replace it.
+       */
+      val pLow = decay(low, 1, lowTimestamp)
+      val pHigh = decay(high, 0, highTimestamp)
+      val x = low + (high - low) * (pLow / (pLow - pHigh))
+      if (amount <= x) {
+        copy(high = amount, highTimestamp = timestamp)
+      } else {
+        copy()
+      }
+    }
+
+  def couldSend(amount: MilliSatoshi, timestamp: TimestampSecond): BalanceEstimate =
+    otherSide.couldNotSend(totalCapacity - amount, timestamp).otherSide
+
+  def didSend(amount: MilliSatoshi, timestamp: TimestampSecond): BalanceEstimate =
+    copy(low = (low - amount) max MilliSatoshi(0), high = (high - amount) max MilliSatoshi(0))
+
+  def addChannel(capacity: Satoshi): BalanceEstimate = copy(high = high + toMilliSatoshi(capacity), totalCapacity = totalCapacity + capacity)
```

Comment:
For a very long time I was in favor of combining parallel channels for Uncertainty Estimates into one large virtual channel with the sum of the capacities. However this becomes really tricky if these channels have different fee policies and one may wish to include those as an additional feature to the cost function of the min cost flow problem. I finally arrived at the point where I maintain those parallel channels in my data structure. While I belief it messes up the uncertainty prediction in particular as node operators may chose to forward on one of the other channels I think it is a "them" problem and not a "me" problem

Reply:
> However this becomes really tricky if these channels have different fee policies

Do note that it doesn't make any sense IMO to have difference fee policies on different channels with the same peer (because non-strict forwarding)